src/pstore/pstore.c

```diff
@@ -0,0 +1,517 @@
+                                        (strcmp(pe->direntbuf.d_name, pe->dll_prev->direntbuf.d_name) > 0)) {
+                                        LIST_REMOVE(dll, (*list), pe);
+                                        LIST_PREPEND(dll, (*list), pe);
+                                        sorted = 0;
+                                        break; /* re-start sort */
+                                }
+                        }
+                }
+        } while (!sorted);
+}
+
+static void free_files(pstore_entry_t **list) {
+        pstore_entry_t *pe;
+        LIST_FOREACH(dll, pe, *list) {
+                free(pe->contents);
+                LIST_REMOVE(dll, (*list), pe);
+                free(pe);
+        }
+}
+
+static int move_file(pstore_entry_t *pe, const char *subdir) {
+        _cleanup_free_ char *ifd_path;
+        int r = 0;
+
+        ifd_path = path_join(arg_sourcedir, pe->direntbuf.d_name);
+        if (!ifd_path)
+                return log_oom();
+
+        if (arg_storage == PSTORE_STORAGE_EXTERNAL) {
+                /* This code copies the file from the pstore to other storage.
+                 * The rename() syscall is not utilized as it results in the
+                 * Invalid cross-device link error.
+                 * In addition, an optional subdirectory can be specified in
+                 * forming the final destination path.  */
+                _cleanup_free_ char *ofd_path;
+                ofd_path = path_join(arg_archivedir, subdir, pe->direntbuf.d_name);
+                if (!ofd_path)
+                        return log_oom();
+
+                /* Make sure destination exists */
+                r = mkdir_parents(ofd_path, 0755);
+                if (r == 0) {
+                        r = copy_file_atomic(ifd_path, ofd_path, 0644, 0, 0, COPY_REPLACE);
+                        if (r != 0)
+                                return log_error("Failed to copy_file_atomic: %s to %s", ifd_path, ofd_path);
+                }
+                else
+                        return log_error_errno(errno, "Failed to create directoy %s: %m", ofd_path);
+        }
+        else
+        if (arg_storage == PSTORE_STORAGE_JOURNAL) {
+                //sd_journal_send("MESSAGE=%s", pe->direntbuf.d_name, NULL);
+                if (pe->is_binary) {
+                        //WITH_BINARY= but how to know end of binary blob if contained in 0 terminated string?
+                }
+                else {
+                        /* NOTE: If journald not running, r always 0 */
+                        r = sd_journal_send("MESSAGE=File %s:\n%s", pe->direntbuf.d_name, pe->contents, NULL);
+                        if (r != 0)
+                                return log_error("Failed to sd_journal_send: %s", ifd_path);
+                }
+        }
+
+        /* If file copied properly, remove it from pstore */
+        if ((r == 0) && arg_unlink)
+                (void)unlink(ifd_path);
```

Comment:
We usually add a white space after cast.
```
(void) func();
```

Reply:
done